File: src/markov/MarkovChainGen2.java
package markov;

import java.util.List;
import java.util.LinkedList;
import java.util.Map;

public class MarkovChainGen2 extends MarkovChain {
    private MarkovWordGraph wgPairs;

    private String secondLastWord;

    public MarkovChainGen2() {
        super();
        wgPairs = new MarkovWordGraph(START_SIGNAL, END_SIGNAL);

        secondLastWord = null;
    }
    public MarkovChainGen2(String filename){
        this();
        train(filename);
    }

    public String getSecondLastWord() {
        return secondLastWord;
    }

    @Override
    public void train(String filename) {
        super.train(filename);

        wgPairs.processFileGen(FILEPATH + filename, 2);
    }

    @Override
    public List<String> getNextWords() {
        var firstOrder = super.getNextWords();

        if (secondLastWord == null) return firstOrder;

        var findInText = "[" + secondLastWord + ", " + getLastWord() + "]";

        return getStrings(firstOrder, findInText, wgPairs);
    }

    static List<String> getStrings(List<String> firstOrder, String findMe, MarkovWordGraph wgPairs) {
        var data = wgPairs.getGraph().getNeighborWeights(findMe);

        List<String> retList = new LinkedList<>();
        if (!data.isEmpty()) {
            for (String word : data.keySet())
                for (int i = 0; i < data.get(word); i++)
                    retList.add(word);
        } else retList = firstOrder;
        return retList;
    }

    @Override
    public void updateMemory() {
        secondLastWord = getLastWord();
    }

    @Override
    public String generateSentence() {
        secondLastWord = null;

        return super.generateSentence();
    }
}

File: src/markov_kt/MarkovChainGen2.kt
package markov_kt

import markov.MarkovChain
import java.util.*

class MarkovChainGen2() : MarkovChain() {
    private val wgPairs: MarkovWordGraph
    var secondLastWord: String?
        private set

    init {
        wgPairs = MarkovWordGraph(START_SIGNAL, END_SIGNAL)
        secondLastWord = null
    }

    constructor(filename: String) : this() {
        train(filename)
    }

    override fun train(filename: String) {
        super.train(filename)
        wgPairs.processFileGen(FILEPATH + filename, 2)
    }

    override fun getNextWords(): List<String> {
        val firstOrder = super.getNextWords()
        if (secondLastWord == null) return firstOrder
        val findInText = "[" + secondLastWord + ", " + getLastWord() + "]"
        return getStrings(firstOrder, findInText, wgPairs)
    }

    override fun updateMemory() {
        secondLastWord = getLastWord()
    }

    override fun generateSentence(): String {
        secondLastWord = null
        return super.generateSentence()
    }

    companion object {
        fun getStrings(firstOrder: MutableList<String>, findMe: String?, wgPairs: MarkovWordGraph): List<String> {
            val data = wgPairs.getGraph().getNeighborWeights(findMe)
            var retList: MutableList<String> = LinkedList()
            if (!data.isEmpty()) {
                for (word in data.keys) for (i in 0 until data[word]!!) retList.add(word)
            } else retList = firstOrder
            return retList
        }
    }
}